### ISSUE TITLE
Intermittent SignatureDoesNotMatch errors

### ISSUE THREAD

**Issue by JohnMaguire** [backblaze-b2-samples/cloudflare-b2#5]
Hi there,

I'm running this code in a Cloudflare Worker and since deploying it I have gotten intermittent 403 errors.

```xml
<Error>
	<Code>SignatureDoesNotMatch</Code>
	<Message>Signature validation failed</Message>
</Error>
```

However, since yesterday it is no longer intermittent.

![image](https://github.com/backblaze-b2-samples/cloudflare-b2/assets/440033/f71c7545-81b8-4328-9869-8704479c1f7b)

I reached out to Backblaze support which asked me to provide the API key credentials... I thought this was a bit odd, but they're read-only so I went ahead and provided them, to which they said they are able to authenticate properly and to reach out to... Veeam? I'm not sure where Veeam is supposed to come into play, but I figured I'd stop by here and see if you have any ideas.

The exact code I'm running is:

```
//
// Proxy Backblaze S3 compatible API requests, sending notifications to a webhook
//
// Adapted from https://github.com/obezuk/worker-signed-s3-template
//
import { AwsClient } from 'aws4fetch'

// These headers appear in the request, but are not passed upstream
const UNSIGNABLE_HEADERS = [
  'x-forwarded-proto',
  'x-real-ip',
]

// How many times to retry a range request where the response is missing content-range
const RANGE_RETRY_ATTEMPTS = 3

// URL needs colon suffix on protocol, and port as a string
const HTTPS_PROTOCOL = 'https:'
const HTTPS_PORT = '443'

// Filter out cf-* and any other headers we don't want to include in the signature
function filterHeaders(headers) {
  return new Headers(
    Array.from(headers.entries()).filter(
      (pair) =>
        !UNSIGNABLE_HEADERS.includes(pair[0]) && !pair[0].startsWith('cf-'),
    ),
  )
}

async function handleRequest(event, client) {
  const request = event.request

  // Only allow GET and HEAD methods
  if (!['GET', 'HEAD'].includes(request.method)) {
    return new Response(null, {
      status: 405,
      statusText: 'Method Not Allowed',
    })
  }

  const url = new URL(request.url)

  // Incoming protocol and port is taken from the worker's environment.
  // Local dev mode uses plain http on 8787, and it's possible to deploy
  // a worker on plain http. B2 only supports https on 443
  url.protocol = HTTPS_PROTOCOL
  url.port = HTTPS_PORT

  // Remove leading slashes from path
  let path = url.pathname.replace(/^\//, '')
  // Remove trailing slashes
  path = path.replace(/\/$/, '')
  // Split the path into segments
  const pathSegments = path.split('/')

  if (ALLOW_LIST_BUCKET !== true) {
    // Don't allow list bucket requests
    if (
      (BUCKET_NAME === '$path' && pathSegments.length < 2) || // https://endpoint/bucket-name/
      (BUCKET_NAME !== '$path' && path.length === 0)
    ) {
      // https://bucket-name.endpoint/ or https://endpoint/
      return new Response(null, {
        status: 404,
        statusText: 'Not Found',
      })
    }
  }

  // Set upstream target hostname.
  switch (BUCKET_NAME) {
    case '$path':
      // Bucket name is initial segment of URL path
      url.hostname = B2_ENDPOINT
      break
    case '$host':
      // Bucket name is initial subdomain of the incoming hostname
      url.hostname = url.hostname.split('.')[0] + '.' + B2_ENDPOINT
      break
    default:
      // Bucket name is specified in the BUCKET_NAME variable
      url.hostname = BUCKET_NAME + '.' + B2_ENDPOINT
      break
  }

  console.log('Headers:', JSON.stringify([...request.headers]))
  // Certain headers, such as x-real-ip, appear in the incoming request but
  // are removed from the outgoing request. If they are in the outgoing
  // signed headers, B2 can't validate the signature.
  const headers = filterHeaders(request.headers)
  console.log('Headers:', JSON.stringify([...headers]))

  // Sign the outgoing request
  const signedRequest = await client.sign(url.toString(), {
    method: request.method,
    headers: headers,
  })

  console.log('URL:', signedRequest.url)

  // For large files, Cloudflare will return the entire file, rather than the requested range
  // So, if there is a range header in the request, check that the response contains the
  // content-range header. If not, abort the request and try again.
  // See https://community.cloudflare.com/t/cloudflare-worker-fetch-ignores-byte-request-range-on-initial-request/395047/4
  if (signedRequest.headers.has('range')) {
    let attempts = RANGE_RETRY_ATTEMPTS
    let response
    do {
      let controller = new AbortController()
      response = await fetch(signedRequest.url, {
        method: signedRequest.method,
        headers: signedRequest.headers,
        signal: controller.signal,
      })
      if (response.headers.has('content-range')) {
        // Only log if it didn't work first time
        if (attempts < RANGE_RETRY_ATTEMPTS) {
          console.log(
            `Retry for ${signedRequest.url} succeeded - response has content-range header`,
          )
        }
        // Break out of loop and return the response
        break
      } else {
        attempts -= 1
        console.error(
          `Range header in request for ${signedRequest.url} but no content-range header in response. Will retry ${attempts} more times`,
        )
        controller.abort()
      }
    } while (response.ok && attempts > 0)

    if (attempts < 0) {
      console.error(
        `Tried range request for ${signedRequest.url} ${RANGE_RETRY_ATTEMPTS} times, but no content-range in response.`,
      )
    }

    // Return whatever response we have rather than an error response
    return response
  }

  // Send the signed request to B2, returning the upstream response
  return fetch(signedRequest)
}

// Extract the region from the endpoint
const endpointRegex = /^s3\.([a-zA-Z0-9-]+)\.backblazeb2\.com$/
const [, aws_region] = B2_ENDPOINT.match(endpointRegex)

// Create an S3 API client that can sign the outgoing request
const client = new AwsClient({
  accessKeyId: B2_APPLICATION_KEY_ID,
  secretAccessKey: B2_APPLICATION_KEY,
  service: 's3',
  region: aws_region,
})

addEventListener('fetch', function (event) {
  event.respondWith(handleRequest(event, client))
})
```

Here is an example log:

```
GET https://i.servepic.com/REDACTED.jpg - Ok @ 10/6/2023, 4:31:07 PM
  (log) Headers: [["accept","text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8"],["accept-encoding","gzip, br"],["accept-language","en-US,en;q=0.5"],["cf-connecting-ip","162.194.213.135"],["cf-ipcountry","US"],["cf-ray","8120aa892846871c"],["cf-visitor","{\"scheme\":\"https\"}"],["connection","Keep-Alive"],["dnt","1"],["host","i.servepic.com"],["sec-fetch-dest","document"],["sec-fetch-mode","navigate"],["sec-fetch-site","cross-site"],["upgrade-insecure-requests","1"],["user-agent","Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/118.0"],["x-forwarded-proto","https"],["x-real-ip","162.194.213.135"]]
  (log) Headers: [["accept","text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8"],["accept-encoding","gzip, br"],["accept-language","en-US,en;q=0.5"],["connection","Keep-Alive"],["dnt","1"],["host","i.servepic.com"],["sec-fetch-dest","document"],["sec-fetch-mode","navigate"],["sec-fetch-site","cross-site"],["upgrade-insecure-requests","1"],["user-agent","Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/118.0"]]
  (log) URL: https://servepic-content.s3.us-west-001.backblazeb2.com/1a1e93f2-a2ea-463b-9a95-9c6df7891f46.jpg
```

In wrangler.toml:

```toml
[vars]
B2_APPLICATION_KEY_ID = "redacted"
B2_ENDPOINT = "s3.us-west-001.backblazeb2.com"
# Set BUCKET_NAME to:
#   "A Backblaze B2 bucket name" - direct all requests to the specified bucket
#   "$path" - use the initial segment in the incoming URL path as the bucket name
#           e.g. https://images.example.com/bucket-name/path/to/object.png
#   "$host" - use the initial subdomain in the hostname as the bucket name
#           e.g. https://bucket-name.images.example.com/path/to/object.png
BUCKET_NAME = "redacted"
```

And finally, I set the secret key via `echo "redacted" | wrangler secret put B2_APPLICATION_KEY`.

Thanks in advance.

**Comment by fourdim** [comment]
Solved by #4.

**Comment by metadaddy** [comment]
It looks like this might be the culprit: [Upcoming Change To Accept Encoding Header](https://community.cloudflare.com/t/upcoming-change-to-accept-encoding-header/517115).

**Comment by metadaddy** [comment]
Closing, since this was fixed by #4.
